File: tests/test_tool_details_simple.c
/*
 * Test: Tool Details Display (Simple)
 * Purpose: Verify MCP tool name extraction with generic parameter handling
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <cjson/cJSON.h>

// Simplified version of the MCP tool details logic (matching new implementation)
static char* get_mcp_tool_details_simple(const char *tool_name, cJSON *arguments) {
    static char details[256];
    details[0] = '\0';

    if (strncmp(tool_name, "mcp_", 4) != 0) {
        return NULL;
    }

    // Extract the actual tool name after the server prefix for display
    const char *actual_tool = strchr(tool_name + 4, '_');
    if (!actual_tool) {
        // Fallback: show the full tool name without "mcp_" prefix
        snprintf(details, sizeof(details), "%s", tool_name + 4);
        details[sizeof(details) - 1] = '\0';
        return details;
    }
    actual_tool++; // Skip the underscore

    // Try to extract the most relevant argument for display
    // Common patterns: url, text, path, element, values, etc.
    cJSON *url = cJSON_GetObjectItem(arguments, "url");
    cJSON *text = cJSON_GetObjectItem(arguments, "text");
    cJSON *path = cJSON_GetObjectItem(arguments, "path");
    cJSON *element = cJSON_GetObjectItem(arguments, "element");

    if (cJSON_IsString(url)) {
        // Tools with URL parameter (navigate, fetch, etc.)
        snprintf(details, sizeof(details), "%s: %s", actual_tool, url->valuestring);
    } else if (cJSON_IsString(text) && strlen(text->valuestring) > 0) {
        // Tools with text parameter (type, search, etc.)
        snprintf(details, sizeof(details), "%s: %.30s%s", actual_tool,
                text->valuestring,
                strlen(text->valuestring) > 30 ? "..." : "");
    } else if (cJSON_IsString(path)) {
        // Tools with path parameter (read, write, etc.)
        snprintf(details, sizeof(details), "%s: %s", actual_tool, path->valuestring);
    } else if (cJSON_IsString(element)) {
        // Tools with element parameter (click, hover, etc.)
        snprintf(details, sizeof(details), "%s: %s", actual_tool, element->valuestring);
    } else {
        // Generic display: just show the tool name
        snprintf(details, sizeof(details), "%s", actual_tool);
    }
    details[sizeof(details) - 1] = '\0';

    return strlen(details) > 0 ? details : NULL;
}

// Test helper
static void assert_contains(const char *result, const char *expected, const char *test_name) {
    if (!result) {
        fprintf(stderr, "✗ FAIL %s: Expected result containing '%s', got NULL\n", test_name, expected);
        exit(1);
    }
    if (strstr(result, expected) == NULL) {
        fprintf(stderr, "✗ FAIL %s: Expected '%s' in result '%s'\n", test_name, expected, result);
        exit(1);
    }
    printf("✓ PASS %s: '%s' found in '%s'\n", test_name, expected, result);
}

// Test Playwright browser tools
static void test_playwright_click(void) {
    cJSON *args = cJSON_CreateObject();
    cJSON_AddStringToObject(args, "element", "Submit button");
    cJSON_AddStringToObject(args, "ref", "button-123");

    char *result = get_mcp_tool_details_simple("mcp_playwright_browser_click", args);
    assert_contains(result, "browser_click", "Playwright browser_click tool name");
    assert_contains(result, "Submit button", "Playwright browser_click with element");

    cJSON_Delete(args);
}

static void test_playwright_type(void) {
    cJSON *args = cJSON_CreateObject();
    cJSON_AddStringToObject(args, "element", "Email input");
    cJSON_AddStringToObject(args, "text", "test@example.com");

    char *result = get_mcp_tool_details_simple("mcp_playwright_browser_type", args);
    assert_contains(result, "browser_type", "Playwright browser_type tool name");
    assert_contains(result, "test@example.com", "Playwright browser_type with text");

    cJSON_Delete(args);
}

static void test_playwright_navigate(void) {
    cJSON *args = cJSON_CreateObject();
    cJSON_AddStringToObject(args, "url", "https://example.com");

    char *result = get_mcp_tool_details_simple("mcp_playwright_browser_navigate", args);
    assert_contains(result, "browser_navigate", "Playwright browser_navigate tool name");
    assert_contains(result, "example.com", "Playwright browser_navigate with URL");

    cJSON_Delete(args);
}

static void test_playwright_snapshot(void) {
    cJSON *args = cJSON_CreateObject();

    char *result = get_mcp_tool_details_simple("mcp_playwright_browser_snapshot", args);
    assert_contains(result, "browser_snapshot", "Playwright browser_snapshot tool name");

    cJSON_Delete(args);
}

// Test generic MCP tools (non-playwright)
static void test_generic_fetch(void) {
    cJSON *args = cJSON_CreateObject();
    cJSON_AddStringToObject(args, "url", "https://api.example.com/data");

    char *result = get_mcp_tool_details_simple("mcp_http_fetch", args);
    assert_contains(result, "fetch", "Generic fetch tool name");
    assert_contains(result, "api.example.com", "Generic fetch with URL");

    cJSON_Delete(args);
}

static void test_generic_search(void) {
    cJSON *args = cJSON_CreateObject();
    cJSON_AddStringToObject(args, "text", "search query");

    char *result = get_mcp_tool_details_simple("mcp_search_query", args);
    assert_contains(result, "query", "Generic search tool name");
    assert_contains(result, "search query", "Generic search with text");

    cJSON_Delete(args);
}

static void test_generic_file_read(void) {
    cJSON *args = cJSON_CreateObject();
    cJSON_AddStringToObject(args, "path", "/path/to/file.txt");

    char *result = get_mcp_tool_details_simple("mcp_fs_read", args);
    assert_contains(result, "read", "Generic file read tool name");
    assert_contains(result, "/path/to/file.txt", "Generic file read with path");

    cJSON_Delete(args);
}

static void test_generic_no_params(void) {
    cJSON *args = cJSON_CreateObject();

    char *result = get_mcp_tool_details_simple("mcp_server_status", args);
    assert_contains(result, "status", "Generic tool with no params");

    cJSON_Delete(args);
}

// Test text truncation
static void test_long_text_truncation(void) {
    cJSON *args = cJSON_CreateObject();
    cJSON_AddStringToObject(args, "text", "This is a very long text that should be truncated when displayed in the UI");

    char *result = get_mcp_tool_details_simple("mcp_example_process", args);
    assert_contains(result, "process", "Long text tool name");
    assert_contains(result, "...", "Long text truncation marker");

    cJSON_Delete(args);
}

// Test tool without server prefix (edge case)
static void test_malformed_tool_name(void) {
    cJSON *args = cJSON_CreateObject();

    char *result = get_mcp_tool_details_simple("mcp_noserver", args);
    // Should return the part after "mcp_"
    assert_contains(result, "noserver", "Malformed tool name fallback");

    cJSON_Delete(args);
}

int main(void) {
    printf("Testing MCP tool details display (generic approach)...\n\n");

    // Test Playwright tools (should work with generic logic)
    printf("Testing Playwright tools:\n");
    test_playwright_click();
    test_playwright_type();
    test_playwright_navigate();
    test_playwright_snapshot();

    // Test generic MCP tools (any server)
    printf("\nTesting generic MCP tools:\n");
    test_generic_fetch();
    test_generic_search();
    test_generic_file_read();
    test_generic_no_params();

    // Test edge cases
    printf("\nTesting edge cases:\n");
    test_long_text_truncation();
    test_malformed_tool_name();

    printf("\n✅ All MCP tool details tests passed!\n");
    printf("   Generic parameter detection works for any MCP server/tool\n");
    return 0;
}

File: zig/tests/test_tool_details_simple.zig
//! tests/test_tool_details_simple.zig — Zig port of tests/test_tool_details_simple.c
//!
//! The C test validated MCP tool display-name extraction logic that parses
//! tool names like "mcp_playwright_browser_click" and formats a human-readable
//! detail string for the TUI.  The logic was a standalone C function that
//! inspected cJSON argument objects.
//!
//! In the Zig implementation the same logic lives inside the MCP and TUI
//! layers.  Rather than duplicating the rendering logic in a test, this file
//! tests the underlying string-manipulation primitives that the Zig port uses
//! to implement the same behaviour.
//!
//! The core invariant being tested is:
//!   - An "mcp_<server>_<tool>" name can be split into server+tool parts
//!   - URL/text/path/element arguments can be extracted from JSON
//!   - Long text values are truncated at 30 characters with "..."
//!   - Tools without recognised args show just the tool name
//!   - Non-mcp names are ignored (return null)

const std = @import("std");

// ---------------------------------------------------------------------------
// Re-implementation of the display-name extraction function (pure Zig)
//
// This mirrors the logic from test_tool_details_simple.c, translated to
// idiomatic Zig so it can be exercised without pulling in cJSON.
// ---------------------------------------------------------------------------

/// Extract MCP tool display details from a tool name and parsed arguments.
/// Returns an owned string (caller frees) or null if the tool is not an MCP
/// tool (does not start with "mcp_").
fn getMcpToolDetails(
    allocator: std.mem.Allocator,
    tool_name: []const u8,
    args: std.json.Value,
) !?[]u8 {
    // Only handle mcp_ prefixed names
    if (!std.mem.startsWith(u8, tool_name, "mcp_")) return null;

    // Strip "mcp_" prefix then split on next underscore to get server + actual
    const after_prefix = tool_name[4..];
    const sep = std.mem.indexOfScalar(u8, after_prefix, '_');
    const actual_tool = if (sep) |s| after_prefix[s + 1 ..] else after_prefix;

    // Extract most relevant argument
    const obj = switch (args) {
        .object => |m| m,
        else => return try allocator.dupe(u8, actual_tool),
    };

    if (obj.get("url")) |url_val| {
        if (url_val == .string) {
            return try std.fmt.allocPrint(allocator, "{s}: {s}", .{ actual_tool, url_val.string });
        }
    }
    if (obj.get("text")) |text_val| {
        if (text_val == .string and text_val.string.len > 0) {
            const text = text_val.string;
            if (text.len > 30) {
                return try std.fmt.allocPrint(allocator, "{s}: {s}...", .{ actual_tool, text[0..30] });
            }
            return try std.fmt.allocPrint(allocator, "{s}: {s}", .{ actual_tool, text });
        }
    }
    if (obj.get("path")) |path_val| {
        if (path_val == .string) {
            return try std.fmt.allocPrint(allocator, "{s}: {s}", .{ actual_tool, path_val.string });
        }
    }
    if (obj.get("element")) |elem_val| {
        if (elem_val == .string) {
            return try std.fmt.allocPrint(allocator, "{s}: {s}", .{ actual_tool, elem_val.string });
        }
    }

    return try allocator.dupe(u8, actual_tool);
}

// ---------------------------------------------------------------------------
// Helper
// ---------------------------------------------------------------------------

fn parseArgs(allocator: std.mem.Allocator, json: []const u8) !std.json.Parsed(std.json.Value) {
    return std.json.parseFromSlice(std.json.Value, allocator, json, .{});
}

// ---------------------------------------------------------------------------
// Tests — Playwright tools
// ---------------------------------------------------------------------------

test "tool_details: mcp_playwright_browser_click uses element arg" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator,
        \\{"element":"Submit button","ref":"button-123"}
    );
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_playwright_browser_click", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "browser_click") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "Submit button") != null);
}

test "tool_details: mcp_playwright_browser_type uses text arg" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator,
        \\{"element":"Email input","text":"test@example.com"}
    );
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_playwright_browser_type", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "browser_type") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "test@example.com") != null);
}

test "tool_details: mcp_playwright_browser_navigate uses url arg" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator,
        \\{"url":"https://example.com"}
    );
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_playwright_browser_navigate", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "browser_navigate") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "example.com") != null);
}

test "tool_details: mcp_playwright_browser_snapshot no args returns tool name" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator, "{}");
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_playwright_browser_snapshot", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "browser_snapshot") != null);
}

// ---------------------------------------------------------------------------
// Tests — Generic MCP tools
// ---------------------------------------------------------------------------

test "tool_details: mcp_http_fetch with url" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator,
        \\{"url":"https://api.example.com/data"}
    );
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_http_fetch", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "fetch") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "api.example.com") != null);
}

test "tool_details: mcp_search_query with text" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator,
        \\{"text":"search query"}
    );
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_search_query", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "query") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "search query") != null);
}

test "tool_details: mcp_fs_read with path" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator,
        \\{"path":"/path/to/file.txt"}
    );
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_fs_read", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "read") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "/path/to/file.txt") != null);
}

test "tool_details: mcp_server_status no params" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator, "{}");
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_server_status", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "status") != null);
}

// ---------------------------------------------------------------------------
// Edge cases
// ---------------------------------------------------------------------------

test "tool_details: long text is truncated with ellipsis" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator,
        \\{"text":"This is a very long text that should be truncated when displayed in the UI"}
    );
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_example_process", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "process") != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "...") != null);
}

test "tool_details: malformed name with no server prefix falls back" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator, "{}");
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "mcp_noserver", parsed.value);
    defer if (result) |r| allocator.free(r);

    try std.testing.expect(result != null);
    try std.testing.expect(std.mem.indexOf(u8, result.?, "noserver") != null);
}

test "tool_details: non-mcp tool returns null" {
    const allocator = std.testing.allocator;
    const parsed = try parseArgs(allocator, "{}");
    defer parsed.deinit();

    const result = try getMcpToolDetails(allocator, "Bash", parsed.value);
    try std.testing.expect(result == null);
}
